tests/integration/ha_tests/test_replication.py

```diff
@@ -0,0 +1,135 @@
+from pytest_operator.plugin import OpsTest
+from tenacity import Retrying, stop_after_delay, wait_fixed
+
+from tests.integration.ha_tests.conftest import APPLICATION_NAME
+from tests.integration.ha_tests.helpers import (
+    are_writes_increasing,
+    check_writes,
+    is_cluster_updated,
+    start_continuous_writes,
+)
+from tests.integration.helpers import (
+    CHARM_SERIES,
+    app_name,
+    build_and_deploy,
+    db_connect,
+    get_password,
+    get_primary,
+    get_unit_address,
+    scale_application,
+)
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest) -> None:
+    """Build and deploy three unit of PostgreSQL."""
+    wait_for_apps = False
+    # It is possible for users to provide their own cluster for HA testing. Hence, check if there
+    # is a pre-existing cluster.
+    if not await app_name(ops_test):
+        wait_for_apps = True
+        await build_and_deploy(ops_test, 3, wait_for_idle=False)
+    # Deploy the continuous writes application charm if it wasn't already deployed.
+    if not await app_name(ops_test, APPLICATION_NAME):
+        wait_for_apps = True
+        async with ops_test.fast_forward():
+            charm = await ops_test.build_charm("tests/integration/ha_tests/application-charm")
+            await ops_test.model.deploy(
+                charm, application_name=APPLICATION_NAME, series=CHARM_SERIES
+            )
+
+    if wait_for_apps:
+        async with ops_test.fast_forward():
+            await ops_test.model.wait_for_idle(status="active", timeout=1000)
+
+
+async def test_reelection(ops_test: OpsTest, continuous_writes, primary_start_timeout) -> None:
+    """Kill primary unit, check reelection."""
+    app = await app_name(ops_test)
+    if len(ops_test.model.applications[app].units) < 2:
+        await scale_application(ops_test, app, 2)
+
+    # Start an application that continuously writes data to the database.
+    await start_continuous_writes(ops_test, app)
+
+    # Kill the current primary.
+    primary_name = await get_primary(ops_test, app)
+    client = Client(namespace=ops_test.model.info.name)
+    client.delete(Pod, name=primary_name.replace("/", "-"))
+
+    # Wait and get the primary again (which can be any unit, including the previous primary).
+    async with ops_test.fast_forward():
+        await ops_test.model.wait_for_idle(apps=[app], status="active")
+
+    # Check whether writes are increasing.
+    await are_writes_increasing(ops_test, primary_name)
+
+    # Verify that a new primary gets elected (ie old primary is secondary).
+    for attempt in Retrying(stop=stop_after_delay(60), wait=wait_fixed(3)):
+        with attempt:
+            new_primary_name = await get_primary(ops_test, app, down_unit=primary_name)
+            assert new_primary_name != primary_name, "primary reelection hasn't happened"
+
+    # Verify that all the units from the first cluster are up-to-date.
+    await is_cluster_updated(ops_test, primary_name)
```

Comment:
`is_cluster_updated` runs `is_secondary_up_to_date`.
I think `is_secondary_up_to_date` should check not only count but also `MAX(number)`

Reply:
Isn't `test_consistency` checking this?

Reply:
but we don't run `test_consistency` in this test 😞

Reply:
Updated on https://github.com/canonical/postgresql-k8s-operator/pull/137/commits/aa3db197bb6a4e89dbd002cf65eadd065787302c.